sys/kern/subr_uio.c

```diff
@@ -452,33 +462,48 @@ struct uio *
 allocuio(u_int iovcnt)
 {
 	struct uio *uio;
-	int iovlen;
 
-	iovlen = iovcnt * sizeof(struct iovec);
-	uio = malloc(iovlen + sizeof(*uio), M_IOV, M_WAITOK);
-	uio->uio_iov = (struct iovec *)(uio + 1);
+	uio = uma_zalloc_arg(uio_zone, (void *)(uintptr_t)(iovcnt), M_WAITOK);
+#if defined(__CHERI_PURE_CAPABILITY__) && defined(__CHERI_SUBOBJECT_BOUNDS__)
+	KASSERT(cheri_getlen(uio->uio_inline_iov) == UIO_INLINE_IOV * sizeof(struct iovec),
+	    ("Malformed UIO structure, uio_inline_iov is not representable"));
+#endif
+	uio->uio_iov = uio->uio_inline_iov;
+	uio->uio_flags = 0;
+	if (iovcnt > UIO_INLINE_IOV) {
```

Comment:
Do you want to further restrict bounds if `iovcnt < UIO_INLINE_IOV`? Hmm, I guess you can't since you'd be modifying the "reference" copy of this capability. Perhaps we should be paranoid and zero out the rest of the inline entries then?

Reply:
I think I can further restrict bounds, that is a good point. The capability in `uio_iov` is considered to be safely mutable as I will detect that this is an internal allocation from the flags.